---
title: "GSE132040 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE132040
Title: 'Expression profiling by high throughput sequencing'

Experiment: 'This dataset contains the transcriptomic information of 17 organs from Mus musculus across the organism's life span.'


Disease models: 
samples from 10 times-points every 3 months in 27 months-long lifespan 

something is wrong!


```{r count table}

library(DESeq2)

#setwd()
GSE_ID <- 'GSE132040'
temp <- read.csv('GSE132040_190214_A00111_0269_AHH3J3DSXX_190214_A00111_0270_BHHMFWDSXX.csv')
annotation <- read.csv('GSE132040_MACA_Bulk_metadata.csv')
annotation <- annotation[grep("Liver", annotation$source.name),]

#change colnames of dataframe so that they match annotations
colnames <- colnames(temp)
colnames <- gsub("\\..*","",colnames)
colnames(temp) <- colnames
# ok so far 
rownames(temp) <- temp$gene
temp <- temp[,annotation$Sample.name] 
write.csv(temp, 'GSE132040_counttable_only_LIVER')


counts <-as.matrix(temp)
rownames(counts) <- rownames(temp)

# BEWARE (!!!)  it is very important to check manually that the columns of the count matrix correspond to the rows of the sample information table.
identical(colnames(counts), annotation$Sample.name, attrib.as.set = FALSE)
remove(temp)

# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ characteristics..age)

male_annotation <- annotation[annotation$characteristics..sex=='m',]
  
ds_m <- DESeqDataSetFromMatrix(
  countData = counts[,male_annotation$Sample.name],
  colData = male_annotation,
  design = ~ characteristics..age
)

levels(ds$characteristics..age) 
# first measurement is first but they are not in order :/ ? 

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]
ds_m <- ds_m[rowSums(counts(ds_m)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)
vsd_m <- vst(ds_m, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))

```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$characteristics..sex, '-', vsd$characteristics..age)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

3. PCA plot

```{r PCA, echo = FALSE}

pcadata <- plotPCA(vsd, intgroup = c('characteristics..sex','characteristics..age'), 
                   returnData = TRUE)
pcadata

library(ggplot2)

percentVar <- round(100 * attr(pcadata, "percentVar"))

ggplot(pcadata, aes(x = PC1, y = PC2, color = characteristics..age, 
                    shape = characteristics..sex, alpha = 0.75)) + 
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data") 

# just by looking at the plot and finding the values 
outliers <- c('F3_384Bulk_Plate1_S123','P13_384Bulk_Plate2_S373',
              'P18_384Bulk_Plate1_S378')

#let's remove these guys and rerun everything
dim(counts)
counts <- counts[, ! colnames(counts) %in% outliers]
#don't forget to change the annotation file to fit the new count table
annotation <- annotation[!annotation$Sample.name %in% outliers , ]

# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ characteristics..age)

levels(ds$characteristics..age) 
# first measurement is first but they are not in order :/ ? 

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))

```
```{r PCA, echo = FALSE}
# new pca and heat-map without outliers 
pcadata <- plotPCA(vsd, intgroup = c('characteristics..sex','characteristics..age'), returnData = TRUE)
percentVar <- round(100 * attr(pcadata, "percentVar"))
ggplot(pcadata, aes(x = PC1, y = PC2, color = characteristics..age, 
                    shape = characteristics..sex)) + 
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data")


```
```{r heatmap, echo=FALSE}
#heatmap after removing outliers
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$characteristics..sex, '-', vsd$characteristics..age)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

## Differential Expression Analysis 

```{r differential expression analysis, echo=TRUE}

#now, we will create subset count table, where will have place only 6,9,12 month old mice and 18,21,24 months old ones. These will be divided in the categories young and oldm, for the differential expression analysis between those two groups
annotation<- annotation[annotation$characteristics..age %in% c('6','9','12', '18','21','24') , ]

annotation$age_group[annotation$characteristics..age=='6'] <- 'young'
annotation$age_group[annotation$characteristics..age=='9'] <- 'young'
annotation$age_group[annotation$characteristics..age=='12'] <- 'young'
annotation$age_group[annotation$characteristics..age=='18'] <- 'old'
annotation$age_group[annotation$characteristics..age=='21'] <- 'old'
annotation$age_group[annotation$characteristics..age=='24'] <- 'young'

counts <- counts[, colnames(counts) %in% annotation$Sample.name]

# create a DESeqDataSet with new count table and different design!
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ age_group)

levels(ds$age_group)
ds$age_group <-relevel(ds$age_group, 'young')
# first measurement is first but they are not in order :/ ? 

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

ds <- DESeq(ds)

print(summary(results(ds)))

res <- results(ds)

```

## Output for liver disease score 

Desired output is a table containing a subset of 200 genes with lowest p-values (not adjusted) and the following columns: (ENSEMBLE Gene IDs), Gene Symbol, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are misssing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}
res <- res[order(res$pvalue),]

output <- data.frame('Gene Symbol' = rownames(res), 
                    'log2FoldChange' = res$log2FoldChange,
                    'pvalue' = res$pvalue,
                    'padj' = res$padj)
write.csv(output, file = paste(GSE_ID, '_young(6_9_12)vs_old(18_21_24)_deg.csv'), 
          row.names = FALSE)
head(res)
remove(list = ls())

```
