---
title: "GSE135050 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE135050

Title: 'RNA sequencing analysis of hepatic transcriptome profiles in experimental nonalcoholic steatohepatitis'

Experiment: 'Two-month-old C57BL/6J male mice were placed on either chow diet or a diet enriched in high fat, cholesterol, and fructose (Research diet D09100301: 40 kcal% fat, 2% cholesterol, 20 kcal% fructose, HFCF diet) for 1 or 3 months. RNA-seq was used to analyze hepatic gene expression from mice on 1-month chow diet, 1-month HFCF diet, 3-month chow diet, and 3-month HFCF.'

Disease model: High fat high cholesterol high fructose diet for 1 or three months 


Two conditions: CHOW and high fat high cholesterol high fructose (HFCF)
Two time points: 1 month and 3 months 
3 replicates per condition and time-point -> total 12 samples 
  WT Chow 1-month 1 [OneMC0]
	WT Chow 1-month 2 [OneMC1]
	WT Chow 1-month 3 [OneMC2]
	WT HFCF 1-month 1 [OneMHF0]
	WT HFCF 1-month 2 [OneMHF1]
	WT HFCF 1-month 3 [OneMHF2]
	WT Chow 3-months 1 [ThreeMC0]
	WT Chow 3-months 2 [ThreeMC1]
	WT Chow 3-months 3 [ThreeMC2]
	WT HFCF 3-months 1 [ThreeMHF0]
	WT HFCF 3-months 2 [ThreeMHF1]
	WT HFCF 3-months 3 [ThreeMHF2]


```{r count table}

library(readxl) 
library(DESeq2)

GSE_ID <- 'GSE135050'
temp <- read_xlsx('GSE135050_1m3mHFCF_processed_changed.xlsx', sheet = 'OnlyCounts')
annotation <- read_xlsx('GSE135050_1m3mHFCF_processed_changed.xlsx', sheet = 'SampleData')
counts <- as.matrix(temp[,4:15])
counts <- as.matrix(counts)
rownames(counts) <- temp$`Gene ID` 

remove(temp)

# create a DESeqDataSet obeject from count matrix and a table of sample information
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
 design = ~ diet)

levels(ds$diet)

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))

```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$diet, vsd$months, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

3. PCA plot

```{r PCA, echo = FALSE}
plotPCA(vsd, intgroup = c('diet', 'months'))

```




## Differential Expression Analysis 

```{r differential expression analysis, echo=TRUE}
ds <- DESeq(ds)

print(summary(results(ds)))

```

```{r deg}
# downregulated_ds <- ds_res[ ds_res$log2FoldChange< 0, ]
# upregulated_ds <- ds_res[ds_res$log2FoldChange > 0,]
# downregulated <- downregulated[order(downregulated$log2FoldChange),]
# upregulated <- upregulated[order(upregulated$log2FoldChange, decreasing = TRUE),]
# print('Most up regulated genes:')
# upregulated
# 
# print('Most down regulated genes:')
# downregulated
```


```{r plotting results, echo=FALSE}
# 
# plotCounts(ds, gene = 'ENSMUSG00000030483_Cyp2b10', intgroup = 'diet')
# plotCounts(ds, gene = 'ENSMUSG00000050423_Ppp1r3g', intgroup = 'diet')
# 
# library('ggplot2')
# library('ggbeeswarm')
# 
# gctemp <- plotCounts(ds, gene = 'ENSMUSG00000050423_Ppp1r3g', intgroup = c('diet','months'), returnData = TRUE)
# 
# ggplot(gctemp, aes(x = diet, y = count, color = months)) + scale_y_log10() + geom_beeswarm( cex = 3)

```

## Output for liver disease score 

Desired output is a table containing a subset of 200 genes with lowest p-values (not adjusted) and the following columns: ENSEMBLE Gene IDs, Gene Symbol, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are missing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}
results <- results(ds)
results <- results[order(results$pvalue),]

output <- DataFrame('Ensemble gene Ids' = sub('_.*','',rownames(results)),
                    'Gene Symbol' = sub('.*_','',rownames(results)), 
                    'log2FoldChange' = results$log2FoldChange,
                    'pvalue' = results$pvalue,
                    'padj' = results$padj)

write.csv(output, file = paste(GSE_ID, '_hfcfdiet_deg.csv'), row.names = FALSE)

```
