---
title: "GSE148849 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE148849
Title: 'Gene Expression Analysis of the response to ACC inhibition'

Experiment: 'Liver samples from mice fed a normal diet or fed an FFD and treated with vehicle or an ACC inhibitor.
Primary hepatic stellate cells from three donors treated with an ACC inhibitor, TGF beta, and/or a TGF inhibitor.'


Disease models: 
Standard Diet (8 samples) vs Fast Food Diet + Vehicle (10 samples)
(the not-vehicle treatment was not aligned because not needed for DEG analysis)

```{r count table}

library(DESeq2)

#setwd()
GSE_ID <- '../GSE148849'
temp <- read.csv('GSE148849_count_table', sep = ' ')
annotation <- read.csv('SraRunTable.txt')

# BEWARE (!!!)  it is very important to check manually that the columns of the count matrix correspond to the rows of the sample information table.

counts <-as.matrix(temp[,c(2:19)])
rownames(counts) <- temp$gene_symbol 
annotation <- annotation[order(annotation$Run),]
annotation <- annotation[-c(19:74),]
remove(temp)
# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ diet )

levels(ds$diet)
ds$diet <- relevel(ds$diet, 'normal chow')

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))
```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- vsd$diet
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

3. PCA plot

```{r PCA, echo = FALSE}
plotPCA(vsd, intgroup = 'diet')
```
## Differential Expression Analysis 

```{r differential expression analysis, echo=TRUE}

ds <- DESeq(ds)

print(summary(results(ds)))

res <- results(ds)

```

## Output for liver disease score 

Desired output is a table containing a subset of 200 genes with lowest p-values (not adjusted) and the following columns: (ENSEMBLE Gene IDs), Gene Symbol, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are misssing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}
res <- res[order(res$pvalue),]

output <- data.frame('Gene Symbol' = rownames(res), 
                    'log2FoldChange' = res$log2FoldChange,
                    'pvalue' = res$pvalue,
                    'padj' = res$padj)
write.csv(output, file = paste(GSE_ID, '_fastfooddiet_deg.csv'), 
          row.names = FALSE)
head(res)
remove(list = ls())

```
