---
title: "GSE138419 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE138419
Title: 'Neutralization of Oxidized Phospholipids Restrains Nonalcoholic Steatohepatitis'

Experiment: 'Using a transgenic mouse that expresses a functional single chain variable fragment of E06, a natural antibody that neutralizes OxPL, we demonstrate the casual role of OxPL in NASH./


Disease models: AMLN diet-fed Ldlr-/- mice
Samples: 
	Mouse_E06Ldlr_M_Liver_RNAseq_PolyA_AMLN_rep1
	Mouse_E06Ldlr_M_Liver_RNAseq_PolyA_AMLN_rep2
	Mouse_E06Ldlr_M_Liver_RNAseq_PolyA_AMLN_rep3
	Mouse_LDLR_M_Liver_RNAseq_PolyA_AMLN_rep1
  Mouse_LDLR_M_Liver_RNAseq_PolyA_AMLN_rep2
	Mouse_E06Ldlr_M_Liver_RNAseq_PolyA_Chow_rep1
  Mouse_E06Ldlr_M_Liver_RNAseq_PolyA_Chow_rep2
	Mouse_E06Ldlr_M_Liver_RNAseq_PolyA_Chow_rep3
	Mouse_LDLR_M_Liver_RNAseq_PolyA_Chow_rep1
	Mouse_LDLR_M_Liver_RNAseq_PolyA_Chow_rep2
	Mouse_LDLR_M_Liver_RNAseq_PolyA_Chow_rep3
	
	-> will only use data from LDLR mice that do not have the E06 fragment: 
	3 control samples (9,10,11) and 2 AMLN samples (4,5)


```{r count table}

library(DESeq2)

#setwd()
GSE_ID <- 'GSE138419'
temp <- read.csv('GSE138419_count_table', sep = ' ')
annotation_temp <- read.csv('SraRunTable.txt')

# BEWARE (!!!)  it is very important to check manually that the columns of the count matrix correspond to the rows of the sample information table.

counts <-as.matrix(temp[,c(5,6,10,11,12)])
rownames(counts) <- temp$gene_symbol 
annotation <- annotation_temp[c(4,5,9,10,11),]

# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ diet )
ds$diet <- relevel(ds$diet, 'Chow')
levels(ds$diet)

#
counts_all <-as.matrix(temp[,-1])
rownames(counts_all) <- temp$gene_symbol 

ds_all <- DESeqDataSetFromMatrix(
  countData = counts_all,
  colData = annotation_temp,
  design = ~ diet + Genotype )

levels(ds_all$diet)
ds_all$diet <- relevel(ds_all$diet, 'Chow')
ds_all$Genotype <- relevel(ds_all$Genotype, 'Ldlr-/-')
ds_all <- ds_all[rowSums(counts(ds_all)) > 1,]
vsd_all <- vst(ds_all, blind = T)
sampleDists_all <- dist(t(assay(vsd_all)))

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))
```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste(vsd$Genotype,"-",vsd$diet)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists_all )
rownames(sampleDistMatrix) <- paste(vsd_all$Genotype,"-",vsd_all$diet)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```


3. PCA plot

```{r PCA, echo = FALSE}
plotPCA(vsd, intgroup = 'diet')
plotPCA(vsd_all, intgroup = c('diet','Genotype'))
```
## Differential Expression Analysis 

```{r differential expression analysis, echo=TRUE}

ds <- DESeq(ds)

print(summary(results(ds)))

res <- results(ds)

```

## Output for liver disease score 

Desired output is a table containing a subset of 200 genes with lowest p-values (not adjusted) and the following columns: (ENSEMBLE Gene IDs), Gene Symbol, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are misssing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}
# experimental design: ~ diet_duration + diet
# and ~ diet_duration + ccl4_treatment
res <- res[order(res$pvalue),]

output <- data.frame('Gene Symbol' = rownames(res), 
                    'log2FoldChange' = res$log2FoldChange,
                    'pvalue' = res$pvalue,
                    'padj' = res$padj)
write.csv(output, file = paste(GSE_ID, '_AMLN_deg.csv'), 
          row.names = FALSE)
head(res)
remove(list = ls())

```
