---
title: "GSE114261 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE114261
Title: '	BET Inhibition Improves NASH and Liver Fibrosis'

Experiment: '	RNA-sequencing of liver tissue from 4 naive vehicle-treated mice, 4 STAM vehicle-treated mice, 4 STAM I-BET151-treated mice'
-> BET treated samples won't be included 

Disease models: STAM mouse NASH model

4 naive vehicle-treated mice vs 4 STAM vehicle treated mice 



```{r count table}

library(DESeq2)

#setwd()
GSE_ID <- 'GSE114261'
temp <- read.csv('GSE114261_count_table', sep = ' ')
annotation <- read.csv('SraRunTable.txt')
annotation <- annotation[-c(9:12),]

counts <-as.matrix(temp[,c(2:9)])
rownames(counts) <- temp$gene_symbol
colnames(counts) <- annotation$Run

# BEWARE (!!!)  it is very important to check manually that the columns of the count matrix correspond to the rows of the sample information table.
identical(colnames(counts), annotation$Run, attrib.as.set = FALSE)
remove(temp)

# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ mouse_model)

levels(ds$mouse_model)
ds$mouse_model <- relevel(ds$mouse_model, 'WT')

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)
#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))

```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$mouse_model)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

3. PCA plot


```{r PCA, echo = FALSE}
# new pca and heat-map without outliers 
plotPCA(vsd, intgroup = 'mouse_model')

```

## Differential Expression Analysis 

```{r differential expression analysis, echo=TRUE}

ds <- DESeq(ds)

print(summary(results(ds)))

res <- results(ds)

```

## Output for liver disease score 

The output is a table containing differentially expressed genes ranked by lowest p-values (not adjusted) and the following columns: ENSEMBLE Gene IDs, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are missing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}
res <- res[order(res$pvalue),]

output <- data.frame('Gene Symbol' = rownames(res), 
                    'log2FoldChange' = res$log2FoldChange,
                    'pvalue' = res$pvalue,
                    'padj' = res$padj)
write.csv(output, file = paste(GSE_ID, '_STAM_deg.csv', sep = ''), 
          row.names = FALSE)
head(res)
remove(list = ls())

```
