---
  title: "GSE123894 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---
  
  ```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE123894
Title: 'Differential Metabolic and Multi-tissue Transcriptomic Responses to Fructose Consumption among Genetically Diverse Mice'

Experiment: 
'To examine inter-individual variability in MetS susceptibility in response to fructose consumption, here we fed three inbred mouse strains, namely C57BL/6J (B6), DBA (DBA) and FVB/NJ (FVB) with 8% fructose in drinking water for 12 weeks. ... We chose 8% fructose to mimic the average fructose consumption found in sugar-sweetened beverages (~10% w/v) consumed in humans.'

*metabolic syndrome (MetS) 
Disease models:
water (control) vs. 8% fructose in water 

--> half of each strain with each treatment
DBA/2 x 36
FVB/NJ x 36 
C57BL/6J x 36




  

```{r count table}

library(DESeq2)

#setwd()
GSE_ID <- 'GSE123894'
temp <- read.csv('GSE123894_count_table', sep = ' ')
# only STANDARD diet needed 
annotation <- read.csv('SraRunTable.txt')


# BEWARE (!!!)  it is very important to check manually that the columns of the count matrix correspond to the rows of the sample information table.

counts <- as.matrix(temp[,c(2:109)])
rownames(counts) <- temp$gene_symbol 
remove(temp)

# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ strain + treatment)

levels(ds$treatment)
ds$treatment <- relevel(ds$treatment, 'Water')

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))
```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste0(vsd$strain, '-', vsd$treatment)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)


```

3. PCA plot
PC1 and PC2 separate the samples almost only by their strain. 

```{r PCA, echo = FALSE}
plotPCA(vsd, intgroup = c('strain','treatment'))

library(ggplot2)

plotPCA_modified <- function(object, intgroup = "condition", ntop = 10000, returnData = FALSE) 
{
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
        length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
        drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = ":"))
    }
    else {
        colData(object)[[intgroup]]
    }
    d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], PC3 = pca$x[, 3], PC4 = pca$x[, 4],  group = group, 
        intgroup.df, name = colnames(object))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[1:4]
        return(d)
    }
}


pcaData <- plotPCA_modified(vsd, intgroup = c('strain','treatment'), returnData = T)
percentVar <- round(100 * attr(pcaData, "percentVar")) 

```
```{r PC2 and PC3, echo = FALSE}
ggplot(pcaData, aes(x = PC2, y = PC3, color = factor(treatment), shape = factor(strain))) +
  geom_point(size =3, aes(fill=factor(treatment))) + 
  xlab(paste0("PC2: ", percentVar[2], "% variance")) + 
  ylab(paste0("PC3: ", percentVar[3], "% variance")) 

```

```{r PC3 and PC4, echo=FALSE}
ggplot(pcaData, aes(x = PC3, y = PC4, color = factor(treatment), shape = factor(strain))) +
  geom_point(size =3, aes(fill=factor(treatment))) + 
  xlab(paste0("PC3: ", percentVar[3], "% variance")) + 
  ylab(paste0("PC4: ", percentVar[4], "% variance")) 
```

PCA shows no separation between different treatment conditions.  So let's look at the samples from each strain separately, there are still quite a lot of them (36) anyways!


```{r echo =FALSE}
## HERE ONLY STRAIN FVB/NJ seems to separate the most on the PCA 
#strains: FVB/NJ
#strains: 'DBA/2J'
#C57BL/6J
counts_fstrain <- counts[,annotation[annotation$strain == 'DBA/2J', 'Run' ]]
annotation_fstrain <- annotation[annotation$strain == 'DBA/2J', ]

ds <- DESeqDataSetFromMatrix(
  countData = counts_fstrain,
  colData = annotation_fstrain,
  design =  ~ treatment
) 

levels(ds$treatment)
ds$treatment <- relevel(ds$treatment, 'Water')

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

plotPCA(vsd, intgroup = c('strain','treatment'))

```

After looking at the three strains separately:only for the DBA/2J strain samples separate clearly between conditions.  (But there is not a proper cluster, and the samples are always in pairs.)

## Differential Expression Analysis 

```{r differential expression analysis, echo=TRUE}

# look at the differentially expressed genes for each state separately 

ds <- DESeq(ds)
print(summary(results(ds)))
res <- results(ds)


```

## Output for liver disease score 

Desired output is a table containing differentially expressed genes ranked by lowest p-values (not adjusted) and the following columns: (ENSEMBLE Gene IDs), Gene Symbol, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are misssing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}

res <- res[order(res$pvalue),]

output <- data.frame('Gene Symbol' = rownames(res), 
                        'log2FoldChange' = res$log2FoldChange,
                        'pvalue' = res$pvalue,
                        'padj' = res$padj)

write.csv(output, file = paste(GSE_ID, '_Fructose_only_DBA2Jstrain_deg.csv'), 
          row.names = FALSE)
```