---
title: "GSE153580 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE153580
Title: 'Integrated gene expression analysis of blood and liver tissue in precancerous condition [mRNA]'

Experiment design: 'Patients with diabetes mellitus (DM) have an epidemiologically higher risk for hepatocellular carcinoma (HCC). In mouse models, administration of streptozotocin (STZ) induces insulin-dependent DM by causing islet beta-cell dysfunction and also induced hepatocellular carcinoma (HCC) after 12 weeks.We attempted to elucidate the carcinogenic mechanism in the precancerous state by using hepatic miRNAs and create precancerous marker using exosomal miRNA. Serum and liver tissues were collected from STZ mice and non-treated mice (CTL mice) at 6, 10 and 12W.'

-> although on one of the descriptions it says 'disease state = hepatocellular carcinoma', the description says the mice get HCC 12 weeks after treatment with STZ and all samples are 12 weeks or younger


Disease models: 

GSM4647591	DIAR control 6w liver 11
GSM4647592	DIAR control 6w liver 12
GSM4647593	DIAR control 6w liver 15
GSM4647594	DIAR control 10w liver 17
GSM4647595	DIAR control 10w liver 19
GSM4647596	DIAR control 10w liver 20
GSM4647597	DIAR control 12w liver 22
GSM4647598	DIAR control 12w liver 23
GSM4647599	DIAR control 12w liver 24

GSM4647600	DIAR STZ 6w liver 26
GSM4647601	DIAR STZ 6w liver 27
GSM4647602	DIAR STZ 6w liver 28
GSM4647603	DIAR STZ 10w liver 32
GSM4647604	DIAR STZ 10w liver 33
GSM4647605	DIAR STZ 10w liver 34
GSM4647606	DIAR STZ 12w liver 36
GSM4647607	DIAR STZ 12w liver 38
GSM4647608	DIAR STZ 12w liver 39



```{r count table}

library(DESeq2)

#setwd()
GSE_ID <- 'GSE153580'
temp <- read.csv('GSE153580_count_table', sep = ' ')
annotation <- read.csv('SraRunTable.txt')

# BEWARE (!!!)  it is very important to check manually that the columns of the count matrix correspond to the rows of the sample information table.

counts <-as.matrix(temp[,c(2:19)])
rownames(counts) <- temp$gene_symbol 
#colnames(counts) <-annotation$GEO_Accession..exp.
remove(temp)

# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ disease_state )

levels(ds$disease_state)
ds$disease_state <- relevel(ds$disease_state, 'normal control')

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists  <- dist(t(assay(vsd)))
```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$disease_state,"-",vsd$time)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```


3. PCA plot

```{r PCA, echo = FALSE}
plotPCA(vsd, intgroup = c('disease_state','time'))

```
## Differential Expression Analysis 

```{r differential expression analysis, echo=TRUE}

ds <- DESeq(ds)

print(summary(results(ds)))
res <- results(ds)

```

## Output for liver disease score 

Desired output is a table containing a subset of 200 genes with lowest p-values (not adjusted) and the following columns: (ENSEMBLE Gene IDs), Gene Symbol, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are missing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}
# experimental design: ~ disease_state
res <- res[order(res$pvalue),]
output <- data.frame('Gene Symbol' = rownames(res), 
                    'log2FoldChange' = res$log2FoldChange,
                    'pvalue' = res$pvalue,
                    'padj' = res$padj)
write.csv(output, file = paste(GSE_ID, '_STZ_deg.csv'), 
          row.names = FALSE)
head(res)
remove(list = ls())

```
