---
title: "GSE83240 - Differentially Expressed Genes - Liver Disease"
author: "Paula Unger Avila"
date: "12/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About data: GSE83240
Title: '	Hepatic transcriptome in CUGBP1 S302A knock-in mice highlights CUGBP1 degradation by Gankyrin through diethylnitrosamine (DEN) activation, leading to fatty liver and liver cancer' 

Experiment design: 
'RNA sequencing was performed on livers from WT and CUGBP1 s302a knock-in mice following treatment with DEN (...)	3-week-old mice (WT and CUGBP1 s302a knock-ins) were injected with DEN at 50ug/g body weight. Livers were collected 30-38 weeks later and stored at -80C.'


Disease models: DEN injection on WT mice vs Control mice
3x3 



```{r count table}

library(DESeq2)

#setwd()
GSE_ID <- 'GSE83240'
temp <- read.csv('GSE83240_count_table', sep = ' ')
annotation <- read.csv('SraRunTable.txt')
annotation <- annotation[annotation$genotype.variation == 'WT',]

# BEWARE (!!!)  it is very important to check manually that the columns of the count matrix correspond to the rows of the sample information table.
counts <- as.matrix(temp[,c(2:7)])
colnames(counts) <- annotation$Run
rownames(counts) <- temp$gene_symbol

# create a DESeqDataSet 
ds <- DESeqDataSetFromMatrix(
  countData = counts,
  colData = annotation,
  design = ~ treatment )

levels(ds$treatment)
ds$treatment <- relevel(ds$treatment, 'None')

# remove rows with zero counts
ds <- ds[rowSums(counts(ds)) > 1,]

#vs transformed counts 
vsd <- vst(ds, blind = T)

#Euclidean distance between samples (they use transformed data)
sampleDists <- dist(t(assay(vsd)))
```

## Exploring Data

1. Heatmap of sample-to-sample distances using the variance stabilizing transformed values 

```{r heatmap, echo=FALSE}
library("pheatmap")
library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- vsd$treatment
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```



```{r pca, echo=FALSE}
plotPCA(vsd_hfd, intgroup = 'treatment')

```

```{r PCA, echo = FALSE}
plotPCA(vsd, intgroup = 'treatment')

```


## Differential Expression Analysis 

```{r differential expression analysis, echo=FALSE}

ds <- DESeq(ds)
print(summary(results(ds)))
res <- results(ds)

```

## Output for liver disease score 

Desired output is a table containing differentially expressed genes ranked by p-values (not adjusted) and the following columns: Gene Symbol, log2FoldChange, p-value and p-adjusted. 

In case the gene symbols are missing in the available count table, the Bioconductor package AnnotationDbi can be used to map the IDs to alternative gene names. 


```{r output, echo=FALSE}
res <- res[order(res$pvalue),]

output <- data.frame('Gene Symbol' = rownames(res), 
                    'log2FoldChange' = res$log2FoldChange,
                    'pvalue' = res$pvalue,
                    'padj' = res$padj)

write.csv(output, file = paste(GSE_ID, '_DEN_deg.csv', sep = ''), 
          row.names = FALSE)

```
